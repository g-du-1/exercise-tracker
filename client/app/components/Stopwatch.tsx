import { useRef, useState } from "react";

export const Stopwatch = () => {
  const [isRunning, setIsRunning] = useState(false);
  const [elapsedTime, setElapsedTime] = useState(0);
  const intervalRef = useRef<NodeJS.Timeout | null>(null);

  const startStopwatch = () => {
    setIsRunning(true);

    intervalRef.current = setInterval(() => {
      setElapsedTime((prevElapsedTime) => prevElapsedTime + 1);
    }, 1000);
  };

  const resetStopwatch = () => {
    if (intervalRef.current) clearInterval(intervalRef.current);

    setElapsedTime(0);
    setIsRunning(false);
  };

  const formatTime = (timeInSeconds: number) => {
    const hours = Math.floor(timeInSeconds / 3600);
    const minutes = Math.floor((timeInSeconds % 3600) / 60);
    const seconds = timeInSeconds % 60;

    return `${hours.toString().padStart(2, "0")}:${minutes
      .toString()
      .padStart(2, "0")}:${seconds.toString().padStart(2, "0")}`;
  };

  return (
    <div>
      <div>{formatTime(elapsedTime)}</div>
      <div>
        <button disabled={isRunning} onClick={startStopwatch}>
          Start
        </button>

        <button disabled={!isRunning} onClick={resetStopwatch}>
          Reset
        </button>
      </div>
    </div>
  );
};
