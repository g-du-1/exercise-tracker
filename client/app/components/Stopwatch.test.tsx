/**
 * @jest-environment jsdom
 */
import { render, fireEvent, screen, act } from "@testing-library/react";
import { Stopwatch } from "./Stopwatch";

jest.useFakeTimers();

describe("Stopwatch", () => {
  it("renders stopwatch with initial time of 00:00:00", () => {
    const result = render(<Stopwatch />);

    expect(result.container).toMatchSnapshot();
  });

  it("starts the stopwatch", () => {
    render(<Stopwatch />);

    const startBtn = screen.getByLabelText("Start Stopwatch");
    fireEvent.click(startBtn);
    expect(startBtn).toBeDisabled();

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(screen.getByText("00:00:05")).toBeInTheDocument();
  });

  it("resets the stopwatch", () => {
    render(<Stopwatch />);

    const resetBtn = screen.getByLabelText("Reset Stopwatch");
    expect(resetBtn).toBeDisabled();

    const startBtn = screen.getByLabelText("Start Stopwatch");
    fireEvent.click(startBtn);
    expect(startBtn).toBeDisabled();

    act(() => {
      jest.advanceTimersByTime(8000);
    });

    fireEvent.click(resetBtn);

    expect(screen.getByText("00:00:00")).toBeInTheDocument();
  });
});
