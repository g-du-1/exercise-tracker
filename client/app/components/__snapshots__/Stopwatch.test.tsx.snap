// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stopwatch renders stopwatch with initial time of 00:00:00 1`] = `
<div>
  <div>
    <div>
      00:00:00
    </div>
    <div>
      <button>
        Start
      </button>
      <button
        disabled=""
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;
